### PR TITLE
feat(manager/mint): add mint support

### LINKED ISSUE #9522: New Manager: Mint
**What would you like Renovate to be able to do?**

I wish it will support [Mint](https://github.com/yonaskolb/Mint).
Mint is a wrapper of Swift Package Manager that specializes in CLI tools.

**Did you already have any implementation ideas?**
Mint provides [Mintfile](https://github.com/yonaskolb/Mint#mintfile) to specify a list of versioned packages.

### PR DESCRIPTION
**Pull request by coffmark**
## Changes

- Add support Mintfile
- Mintfile format is based on https://github.com/yonaskolb/Mint/tree/master/Tests/Fixtures

## Context

- Closes #9522

## Documentation (please check one with an [x])

- [x] I have updated the documentation, `lib/modules/manager/mint/readme.md`
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [x] Both unit tests + ran on a real repository

The following was created pr by renovate
https://github.com/coffmark/renovate-mint-sample/pull/2
https://github.com/coffmark/renovate-mint-sample/pull/3

<details>
<summary>execution log by renovate</summary>

```
$ node -r ts-node/register/transpile-only -- lib/renovate.ts coffmark/renovate-mint-sample --token=${RENOVATE_TOKEN}
 INFO: Repository started (repository=coffmark/renovate-mint-sample)
       "renovateVersion": "0.0.0-semantic-release"
 INFO: Dependency extraction complete (repository=coffmark/renovate-mint-sample, baseBranch=master)
       "stats": {
         "managers": {"mint": {"fileCount": 1, "depCount": 3}},
         "total": {"fileCount": 1, "depCount": 3}
       }
 INFO: Branch created (repository=coffmark/renovate-mint-sample, branch=renovate/swiftgen-swiftgen-6.x)
       "commitSha": "9d2d1378c963191d4e9d85c8db858874367ecbfc"
 INFO: PR created (repository=coffmark/renovate-mint-sample, branch=renovate/swiftgen-swiftgen-6.x)
       "pr": 2,
       "prTitle": "Update dependency SwiftGen/SwiftGen to v6.6.2"
 INFO: Branch created (repository=coffmark/renovate-mint-sample, branch=renovate/yonaskolb-xcodegen-2.x)
       "commitSha": "231205399e3a4db001896e1c45608059cf7414f5"
 INFO: PR created (repository=coffmark/renovate-mint-sample, branch=renovate/yonaskolb-xcodegen-2.x)
       "pr": 3,
       "prTitle": "Update dependency yonaskolb/xcodegen to v2.32.0"
 INFO: Deleting orphan branch (repository=coffmark/renovate-mint-sample, branch=renovate/configure)
 INFO: Repository finished (repository=coffmark/renovate-mint-sample)
       "durationMs": 12177
✨  Done in 20.62s.

```
</details>

Looking forward to hearing back from you! Thank you!

### CODE REVIEW

File: lib/modules/manager/mint/__fixtures__/complex.Mintfile
Comment (HonkingGoose) [inline lib/modules/manager/mint/__fixtures__/complex.Mintfile:3]: Line ending problem on this line.

File: lib/modules/manager/mint/__fixtures__/simple.Mintfile
Comment (HonkingGoose) [inline lib/modules/manager/mint/__fixtures__/simple.Mintfile:3]: Line ending problem on this line.

File: lib/modules/manager/mint/__fixtures__/noVersion.Mintfile
Comment (HonkingGoose) [inline lib/modules/manager/mint/__fixtures__/noVersion.Mintfile:2]: Line ending problem on this line.

File: lib/modules/manager/mint/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mint/readme.md:1]:
```suggestion
Renovate supports updating Mintfiles.
```

This follows the same pattern as the other manager readmes.

File: lib/modules/manager/mint/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mint/readme.md:3]:
```suggestion
[`yonaskolb/Mint` repository on GitHub](https://github.com/yonaskolb/Mint)
```

Use a more descriptive link name.

File: lib/modules/manager/mint/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mint/readme.md:6]:
```suggestion
```

Drop the `**note**` part, because the readme is short enough for users to read it in one go fully.

File: lib/modules/manager/mint/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mint/readme.md:7]:
```suggestion
You must put the library version in the Mintfile:
```

If users must do something, use the word `must`. 😉

File: lib/modules/manager/mint/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mint/readme.md:10]:
```suggestion
// Good:
```

File: lib/modules/manager/mint/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mint/readme.md:14]:
```suggestion
// Bad:
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by HonkingGoose** [comment]
Hi @coffmark, did you know that you can "batch" all the suggestions from a PR into a single commit? That is cleaner than making six new commits. 🙈 

Read the [GitHub Docs about applying suggested changes](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/incorporating-feedback-in-your-pull-request#applying-suggested-changes) to learn more. 😉

**Comment by coffmark** [comment]
Hi, @HonkingGoose! 
Thank you very much! I didn't know that!

File: lib/modules/manager/mint/__fixtures__/complex.Mintfile
Comment (viceice) [inline lib/modules/manager/mint/__fixtures__/complex.Mintfile:1]: please inline those small fixtures

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:31]:
```suggestion
      expect(res).toEqual({ deps: [
        {
          depName: 'SwiftGen/SwiftGen',
          currentValue: '6.6.1',
          datasource: 'git-tags',
          packageName: 'https://github.com/SwiftGen/SwiftGen.git',
        },
        {
          depName: 'yonaskolb/xcodegen',
          currentValue: '2.30.0',
          datasource: 'git-tags',
          packageName: 'https://github.com/yonaskolb/xcodegen.git',
        },
        {
          depName: 'realm/SwiftLint',
          currentValue: '0.48.0',
          datasource: 'git-tags',
          packageName: 'https://github.com/realm/SwiftLint.git',
        },
      ]});
```

simplify 😉

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:41]: see other comment

File: lib/modules/manager/mint/extract.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.ts:8]: You need to handle comments `#`, they can be at line start or after a package.

 https://github.com/yonaskolb/Mint/blob/2f34a6d72c7d3543ef4a7f8f6f950c72eb32cce6/Sources/MintKit/Mintfile.swift#L20-L32

---

extract deps without version and add `skipReason`

https://github.com/renovatebot/renovate/blob/e79402d5a4cf5fdd8af0fbb98b71b80515aae107/lib/modules/manager/ansible-galaxy/collections.ts#L122-L124

File: lib/modules/manager/mint/index.ts
Comment (viceice) [inline lib/modules/manager/mint/index.ts:11]:
```suggestion
  fileMatch: ['(^|\\/)Mintfile$'],
```

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:2]:
```suggestion
import { extractPackageFile } from '.';
```

**Review by viceice — CHANGES_REQUESTED** [review]

File: lib/modules/manager/mint/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mint/readme.md:10]: > realm/SwiftLint @ 0.48.0

Are you sure that you can put spaces between the package name and the version number? It looks wrong to me, but maybe this is allowed syntax in a Mintfile. 😄

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
The test suite complains about not meeting the required coverage.

I also have a question about the docs. 😉

File: lib/modules/manager/mint/__fixtures__/noVersion.Mintfile
Comment (viceice) [inline lib/modules/manager/mint/__fixtures__/noVersion.Mintfile:1]: please inline fixtures

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:16]:
```suggestion
realm/SwiftLint@0.48.0 #commented out
```
looks like a better test

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:21]:
```suggestion
    it('Mintfile With Version Description', () => {
```
check other tests too. 😉

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:106]: why only the first has datasource?

File: lib/modules/manager/mint/extract.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.ts:7]:
```suggestion
  content.split('\n').map(s => s.trim()).forEach((line) => {
```
use `for...of` loop instead `forEach`

File: lib/modules/manager/mint/extract.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.ts:21]: should no longer required with trim above

File: lib/modules/manager/mint/readme.md
Comment (coffmark) [reply lib/modules/manager/mint/readme.md:10]: Hi! Sorry for the late reply 🙇‍♂️
As you say, Mintfile allows its syntax.
Here is the test case. (https://github.com/yonaskolb/Mint/blob/master/Tests/Fixtures/Mintfile#L4)

File: lib/modules/manager/mint/extract.spec.ts
Comment (coffmark) [reply lib/modules/manager/mint/extract.spec.ts]: Sorry!🙇‍♂️ I forgot to add!
I add in https://github.com/renovatebot/renovate/pull/17351/commits/8d465c705df36b735525e659fbcd7cf9252f9273

File: lib/modules/manager/mint/__fixtures__/Mintfile
Comment (HonkingGoose) [inline lib/modules/manager/mint/__fixtures__/Mintfile:3]: I'm curious, why is this testcase not in the fixture itself? It seem like the fixture should be the place to put all your testcases?

```
realm/SwiftLint @ 0.48.0
```

I see you test this line in the `extract.spec.ts` file though:

```ts
const complexMintFileContent = `
SwiftGen/SwiftGen@6.6.1
yonaskolb/xcodegen
realm/SwiftLint @ 0.48.0`;
```

File: lib/modules/manager/mint/extract.spec.ts
Comment (HonkingGoose) [inline lib/modules/manager/mint/extract.spec.ts:16]:
```suggestion
const includesCommentedOutMintFileContent = `
```

May be good to rename this `const`? You should do it in VSCode or a similar editor, so you change all instances.

File: lib/modules/manager/mint/readme.md
Comment (HonkingGoose) [inline lib/modules/manager/mint/readme.md:3]:
```suggestion
Go to the [`yonaskolb/Mint` repository on GitHub](https://github.com/yonaskolb/Mint) to learn more about the Mint package manager.
```

I think this is clearer about what you'll see when you click the link, and why you may want to go to the official repository. 😄

File: lib/modules/manager/mint/__fixtures__/Mintfile
Comment (coffmark) [reply lib/modules/manager/mint/__fixtures__/Mintfile]: Thanks! 😄 I added all cases including comment outs!
https://github.com/renovatebot/renovate/pull/17351/commits/b8482afb46811642d60e3feb7a78b91ef6723537

**Review by HonkingGoose — DISMISSED** [review]
I'm happy with the docs. I'll let the maintainers review the code and tests.

File: lib/modules/manager/mint/__fixtures__/Mintfile
Comment (viceice) [inline lib/modules/manager/mint/__fixtures__/Mintfile:1]: why this still isn't inlined?

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:6]: inline those if they are only used in one test.

File: lib/modules/manager/mint/extract.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.ts:7]: use newline regex from`util/regex` to split lines

File: lib/modules/manager/mint/extract.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.ts:36]: remove, unused when skip reason is set

File: lib/modules/manager/mint/__fixtures__/Mintfile
Comment (coffmark) [reply lib/modules/manager/mint/__fixtures__/Mintfile]: I'm so sorry! Thanks for all your reviews! I implemented with this commit https://github.com/renovatebot/renovate/pull/17351/commits/a3c052fbb9a3923022b97de7c93e31fc8ae5b53c

**Review by rarkins — DISMISSED** [review]

File: lib/modules/manager/mint/extract.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.ts:34]:
```suggestion
```

we don't need this

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:61]:
```suggestion
```

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:66]:
```suggestion
```

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:85]:
```suggestion
```

File: lib/modules/manager/mint/extract.spec.ts
Comment (viceice) [inline lib/modules/manager/mint/extract.spec.ts:110]:
```suggestion
```

**Review by viceice — APPROVED** [review]

**Pull request merged**
